runtime/v2/README.md

```diff
@@ -1,29 +1,174 @@
+a process that on Linux normally
+involves a `fork/exec`. Rather, containerd exposes its network communications port, either
+gRPC or its own leaner [ttRPC](https://github.com/containerd/ttrpc), and expects the runtime
+shim to communicate with containerd over that port to receive all commands, and then route the operations
+to the runtime engine. In this manner, the shim becomes the parent of the container(s).
+
+Since the runtime engine normally is a simple program that does not run long-term as a daemon, but instead
+is invoked as once-and-exit for each command, most container runtime architecture models
+are implemented in two parts:
+
+1. The heavy client like kubelet/docker/nerdctl/podman/rancher
+1. The container runtime manager daemon, like containerd/cri-o/dockerd
+1. The runtime shim
+1. The runtime engine
+
+In some cases, various components of the architecture are combined into a single executable or process, for example,
+shim+engine.
+
+#### runtime shim
+
+The runtime shim is what actually is invoked by containerd. It has minimal options on start beyond
+being provided the communications port for containerd and some bare minimum configuration
+information.
+
+The runtime shim communicates with containerd over ttrpc, and invokes the actual runtime
+via fork/exec.
+
+In theory, it is possible for the shim and the runtime to be the same binary and a single process,
+handling the container in separate threads, but in practice, they usually are kept separate to integrate
+with distinct OCI `runc`-compatible runtime engines.
+
+containerd passes options to the shim over the ttrpc connection, including which runtime binary
+to invoke. These are the `options` for the [`CreateTaskRequest`](https://github.com/containerd/containerd/blob/main/runtime/v2/README.md#container-level-shim-configuration).
+
+#### runtime engine
+
+The runtime engine itself is what actually starts and stops the container, for example [runc](https://github.com/opencontainers/runc).
+
+### Relationship
+
+Since each shim instance communicates with containerd as a daemon, while parenting containers via invoking independent runtimes,
+it is possible to have one shim for multiple containers and invocations. For example,
+you could have one `containerd-shim-runc-v2` communicating with one containerd, and it can
+invoke ten distinct containers.
+
+It even is possible to have one shim for multiple containers, each with its own actual runtime,
+since, as described above, the runtime binary is passed as one of the options in `CreateTaskRequest`.
+
+Practically, containerd does not work this way. Instead, it invokes a new shim for each container,
+leading to a 1:1 relationship between shim and container.
+
+In the case where containers are grouped together, like in a pod or sandbox, containerd will
+invoke one shim per pod or sandbox, even though there are multiple individual containers.
```

Comment:
Ah, so it is the _shim_ that makes the decision.

Let's see if I got this:

1. containerd receives request to create a container
2. containerd launches a shim, which listens on a socket, reports address, exits
3. containerd tells shim over ttrpc, "launch container A"
4. containerd receive request for another container
5. containerd launches a shim, which reports address and exits. **This might be the same address as before or a new one, totally up to the shim; containerd does not care.**
6. containerd tells shim over ttrpc, "launch container B"

Is that it?

Reply:
Yeah, I just reread your previous comment, and that sounds right. Will try to work it in.

Reply:
Step 2 & 5 include the container configuration, step 3 & 6 end up being for controlling the specific container's lifecycle.  That allows the shim to decide whether to have a single socket (and single shim) or multiple sockets (which could be multiple shims or a single shim listening on multiple sockets, though I haven't seen one that does that pattern in practice yet).

Reply:
Updated slightly.